ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/RequiredPluginsInitializer.java

```diff
@@ -36,6 +48,48 @@ public class RequiredPluginsInitializer extends ClasspathContainerInitializer {
 
 	@Override
 	public void initialize(IPath containerPath, IJavaProject javaProject) throws CoreException {
+		if ("main".equals(Thread.currentThread().getName())) { //$NON-NLS-1$
+			// See https://github.com/eclipse-pde/eclipse.pde/issues/1481
+			if (WARNING_LOGGED.compareAndSet(false, true)) {
+				ILog.get().warn(
+						"RequiredPluginsInitializer called from within the UI thread this will badly impact your IDE performance!", //$NON-NLS-1$
+						new RuntimeException("Called from main thread here")); //$NON-NLS-1$
+			}
+			JOB_MAP.compute(javaProject, (jp, oldjob) -> {
+				if (oldjob != null) {
+					oldjob.cancel();
+				}
+				Job job = Job.create(PDECoreMessages.PluginModelManager_InitializingPluginModels, m -> {
+					if (oldjob != null) {
+						try {
+							oldjob.join();
+						} catch (InterruptedException e) {
+						}
+					}
+					setClasspath(jp);
+				});
+				job.addJobChangeListener(new JobChangeAdapter() {
+					@Override
+					public void done(IJobChangeEvent event) {
+						JOB_MAP.remove(jp);
+					}
+				});
+				job.schedule();
+				return job;
+			});
+			return;
+		}
+		Job job = JOB_MAP.get(javaProject);
+		if (job != null) {
+			try {
+				job.join();
```

Comment:
Doesn't this have the effect that the method blocks again until the classpath-set returns?
How would this then improve the start-up performance? Or is it just done to resolve dead-locks?

Reply:
This codepath is only executed when it is called outside the UI (e.g. in a build thread) where it is fine to block, the goal is to join on a possible execution that previously was scheduled in an UI thread.